### PR TITLE
Sex prediction from RNASeq

### PR DESCRIPTION
**Pull request by bill-amadio**
#### Purpose/implementation

_What scientific question is your analysis addressing? 
What was your approach?
If this is not adding an analysis, describe your changes in this section._

Which metadata attributes can we reliably predict from gene expression?  In this analysis, reported_sex is the metadata attribute used as the target variable for prediction from a data set of RNASeq transcripts.  The predictive model is built by elastic-net logistic regression using glmnet and glmnetUtils.

#### Issue

_What GitHub issue does your pull request address?_

84: RNA Expression-based Prediction of Sex

#### Directions for reviewers

_Tell potential reviewers what kind of feedback you are soliciting. 
Are there particular areas that need a closer look?
Is there something you want to discuss further?
Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?_

Accuracy and completeness of the pre-processing.  Highest 25% median absolute deviation samples were used.  This percentage was the first I tried.  Do we need to review higher/lower options?

Can we generalize the selection of samples with reported_gender == male or female only?  I think the manual processing I did can be automated.

Have we extracted all we can from the sex.cva object?

Follow up?  Any of Casey's suggestions seem to be within reach.  What would be most helpful?

#### Results

_If your pull request includes code that produces scientific results, please summarize the results here.
This can help facilitate discussion around interpretation.
Please state what kinds of results are included (e.g., table, figure)._

This code builds a predictive model for reported_gender using the top 25% mean absolute deviation transcripts out of the current kallisto file. The data set is split 70/30 train/test.  On various test runs, accuracy on the holdout was between 96 and 98%. The elastic net alpha on these runs was between 0.34 and 1.  With alpha = 0.34, we get about 200 non-zero coefficients; with alpha = 1, less than 100.

Here are two plots from a recent run.  Optimal alpha = 1.  75 non-zero coefficients.  Test set accuracy = 97.7%.
plot_cva

minlossplot

Do you want the model object?  It is 111 Mb.

#### Docker and continuous integration

_Check all those that apply or remove this section if it is not applicable._

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

### CODE REVIEW

File: analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+best_fit <- sex.cva$modlist[[best_alpha_index]]
+
+#look at best_fit$lambda.min to find best lambda corresponding to tuned alpha
+
+non_zero_features <- which(coef(best_fit, s = best_fit$lambda.min) != 0)
+
+```
+```{r}
+
+# look at coef(best_fit, s=best_fit$lambda.min)[non_zero_features] for coefficients
```
Comment (jaclyn-taroni) [inline analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd]: Is this chunk and the one below it intentionally left empty for discussion?

File: analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+```
+
+
+Build predictive model for reported_gender_response.
+Hold out 30% of the samples to test accuracy after cross-validation tuning of 
+  lambda and alpha regularization parameters.
+
+```{r}
+
+library(glmnet)
```
Comment (jaclyn-taroni) [inline analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd]: Can you put all the `library` calls in a single chunk up at the top of the file please?

**Comment by jaclyn-taroni** [comment]
As of https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/123/commits/d4e9671d340a99c46fcaa0238a03f84caa0fd86c, the dependencies have been added to the Dockerfile and this has been added to continuous integration. I'm going to check those both off.

File: analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+missing_reported_gender_samples <- hist[which(hist[, "reported_gender"] == ""), "Kids_First_Biospecimen_ID"]
+
+```
+
+Calculate the median absolute deviation for each transcript.
+
+```{r}
+library(stats)
+
+mads <- apply(ge[ , 3:1030], 1, function(x) mad(x, high=TRUE))
```
[the remark this answers is not in the document]
Comment (jaclyn-taroni) [reply analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd]: Hardcoding also will not work for small files that we use for testing for continuous integration. That's why we see this failure here: https://circleci.com/gh/AlexsLemonade/OpenPBTA-analysis/388

**Comment by bill-amadio** [comment]
I made all the changes ackie suggested for the previous pull request.

This code uses the kallisto.stranded gene expression file as input and reported gender as the target variable.

The model is built and tested on a 70/30 partition.  set.seed(36354).
  
Filtering to the top 25% highest median absolute deviation transcripts, this run of the code returned 181 transcripts with non-zero coefficients and achieved 95.5% accuracy on the holdout set.

I also run the full set of stranded transcripts: 230 non-zero transcripts, 95.2 accuracy on the holdout set.

For the polya set, the top 25% returned 12 features and 83.3% on the holdout set.  The full polya set
returned 20 features and 83.3% accuracy on the holdout.

These models confirmed the difference between the stranded and polya sets.  I used the top 25% stranded model to predict calls from the polya gene expression data.  The accuracy was 51%.

The full stranded model was slightly better: 62% accuracy.

The 12 non-zero transcripts from the polya top 25% model and the 181 non-zero transcripts from the 
stranded top 25% model have just two transcripts in common: 233.9_ARF5-201 and 430580.6_NBPF1-203.

I looked into Jackie's comment about filtering to top 25% uncovering stranded/pola differences by
intersecting the Top 25% of the V4 (combined) file with the Top 25% of the V5 polya file.  The
files have 72% (36055/50101) of their transcripts in common.  The Top 25% of the V4 file has
98% (49328/50101) of its transcripts in common with the V5 stranded file.  

I have a few proposals for the next version of this code.  Since we now have the library calls
in the opening chunk, why not use that chunk to specify all user options?  We could place assignment
statements for the train/test seed, train/test partition percentages, the median absolute deviation cutoff and the gene expression input file.

Also, the current code makes an unnecessary pass over the histologies file to 1) remove "" values from reported_gender and 2) remove "Not available" values from reported_gender.  I should have done that in one pass.

FYI, I tried to run this code on the rsem-fpkm files, but these files have just one identification
column whereas the kallisto files have two.  We could put a conditional statement, based upon the input file name, in the section where we trim those columns.  Are we looking for one code file to use all available gene expression input files?

stranded_top25_model_plot
stranded_top25_model_minloss

**Comment by jaclyn-taroni** [comment]
Thanks @bill-amadio - I am going to look at this today. Right now I am going to resolve the conflicts in the `Dockerfile` and `.circleci/config.yml`.

**Comment by jaclyn-taroni** [comment]
@bill-amadio you have two titles in you Rmd file, which is why the check is failing. I will edit the Rmd file to resolve this.

File: analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+```{r}
+
+train_percent <- 0.70
+set.seed(36354)
+
+train_set <- sample(1:nrow(df), floor(train_percent*nrow(df)))
+test_set <- setdiff(1:nrow(df), train_set)
+
+ptm <- proc.time()
+sex.cva <- cva.glmnet(df[train_set, ], reported_gender_response[train_set], standardize=TRUE,
```
Comment (jaclyn-taroni) [inline analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd]: It looks like the default for the `nfolds` argument of `cva.glmnet` is `10`. Mostly noting it here for the sake of documentation.

File: analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  tested for the corresponding alpha value.
+
+The best performing lambda value is stored within the modlist object as lambda.min.
+
+We carry forward the predictors with non-zero coefficients in the optimal (minimum CVloss) model.
+
+```{r}
+best_fit <- sex.cva$modlist[[best_alpha_index]]
+
+#look at best_fit$lambda.min to find best lambda corresponding to tuned alpha
```
Comment (jaclyn-taroni) [inline analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd]: I always thought the rule of thumb was to use `lambda.1se` rather than `lambda.min`: https://stats.stackexchange.com/questions/138569/why-is-lambda-within-one-standard-error-from-the-minimum-is-a-recommended-valu

**Comment by jaclyn-taroni** [comment]
Between what you have brought up in your comments here @bill-amadio and what @cgreene laid out in #84, there are a number of things that are not reflected in this notebook that we probably want to do and record in this repository. Specifically, we likely want the flexibility to do the following:

* Train on `reported_gender` labels and evaluate on `germline_sex_estimate` labels
* Train on one selection strategy (e.g., stranded) and evaluate on the other (poly-A)

And maybe any combination of the above. So in essence, we need code that will be general enough to take any matrix of features and labels for both training and evaluation. I'm going to propose that we split up _kinds of steps_ in this notebook into three scripts:

```
01-clean_split_data.R
02-train_elasticnet.R
03-evaluate_model.R
```

I'll detail what I'm thinking about regarding the functionality, inputs, and outputs of each script below. Some steps could go in one of two scripts, looking forward to having a discussion about this structure.

###  `01-clean_split_data.R`

#### Functionality/purpose:

This will clean the gene expression data -- i.e., it drops anything with missing labels in either `reported_gender` or `germline_sex_estimate` and if the training partition size < 1.0, it will split the data into training and testing sets that can be saved as separate files to be used downstream.

#### Inputs and arguments:

* gene expression matrix
* histologies file
* seed with some default
* training partition size with some default
* output filenames or output directory, depending on how it is implemented

#### Outputs:

* Always saves a gene expression matrix that contains the training data to file
* If training partition size < 1.0, a gene expression matrix with the testing data also needs to be saved
* Possibly a TSV file that contains the biospecimen ID, `reported_gender`, and `germline_sex_estimate` information that has been cleaned. I say possibly because I think we'll want the downstream scripts to be robust to "extra" samples, etc.

#### Notes:

* I think you can essentially hard code the `reported_gender` and `germline_sex_estimate` part at this time.
* The output should probably go into the `scratch` directory, but I wouldn't hard code that part.


### `02-train_elasticnet.R`

#### Functionality/purpose:

This script will contain the elastic net model training step, with an option for filtering based on MAD first, and save the model object (what is returned by `cva.glmnet`).

#### Inputs and arguments:

* gene expression matrix for training
* labels, perhaps in the form of a TSV file or the entire histologies TSV where you specify the variable (column) that contains the true labels
* MAD percentile with some default
* seed with some default (req'd for CV folds reproducibility)
* output filename or output directory, depending on how it is implemented

#### Outputs:

`cva.glmnet` object saved with `saveRDS`. You might also want to save diagnostic plots as PNGs  and possibly the features with non-zero coefficients in some tabular format at this step (or alternatively, in `03-evaluate_model.R`).

#### Notes:

I would probably stick the RDS file mentioned above in a directory called `models` within `sex-prediction-from-RNASeq` and have git ignore the models folder by adding a `.gitignore` file to the `sex-prediction-from-RNASeq` folder with the following contents:

```
# models are too large to be tracked and pushed to GitHub
models
```

Here is an example of an analysis-specific `.gitignore` file: https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/02b35888ea808663b554e2a0875a40c40ae73ab8/analyses/transcriptomic-dimension-reduction/.gitignore

### `03-evaluate_model.R`

This is probably the least well-formed in my mind at the moment...

#### Functionality/purpose:

This script will evaluate the model (e.g., the equivalent of `best_fit` here) in terms of accuracy and/or another selected metric on the test dataset.

#### Inputs and arguments:

* model file
* gene expression matrix for testing
* labels, perhaps in the form of a TSV file or the entire histologies TSV where you specify the variable (column) that contains the true labels

#### Outputs:

The metric of interest -- it may be useful to employ [`caret::confusionMatrix`](https://www.rdocumentation.org/packages/caret/versions/6.0-84/topics/confusionMatrix) here and save that output with `saveRDS`. You could then extract whatever metric you want for display (e.g., in a table for multiple models) later on. You could also save the predicted labels in a tabular format. Saving the predicted labels this way will allow you to do things like look at accuracy within a particular histology in a downstream notebook or script.

Save diagnostic plots as PNGs and possibly the features with non-zero coefficients in some tabular format at this step if not done beforehand.

File: analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd [hunk 6]
```diff
@@ -0,0 +1,167 @@
+---
+title: :"Sex Prediction from RNA-Seq"
+author: Bill Amadio
+date: 2019
+---
+Library calls
+```{r}
```
[the remark this answers is not in the document]
Comment (bill-amadio) [reply analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd]: Done.  Ready for add, commit.

File: analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+
+Let's try a model with the top 25% highest mads.
+
+```{r}
+tail_percent <- 0.75
+
+# dropping the gene and transcript identifiers and creating a matrix
+
+gene_expression_mat <- as.matrix(ge[ , -c(1,2)])
```
[the remark this answers is not in the document]
Comment (bill-amadio) [reply analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd]: Using scripts will be fine.

I brought the transcriptomic-dimension-reduction method into the notebook.  It is ready for add, commit.

File: analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+Hold out 30% of the samples to test accuracy after cross-validation tuning of 
+  lambda and alpha regularization parameters.
+
+```{r}
+
+library(glmnet)
+library(glmnetUtils)
+
+train_percent <- 0.70
+train_set <- sample(1:nrow(df), floor(train_percent*nrow(df)))
```
[the remark this answers is not in the document]
Comment (bill-amadio) [reply analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd]: Done and ready for add, commit.

File: analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd (hunk 4 above)
Comment (bill-amadio) [reply analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd]: OK.

File: analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd (hunk 5 above)
Comment (bill-amadio) [reply analyses/sex-prediction-from-RNASeq/sex-prediction-from-RNASeq.Rmd]: Ugh!  Thank you for catching this.  When we choose s = lambda.1se we get far fewer non-zero features (113 v.181) and the same predictive accuracy (95.8 v. 95.5).

lambda.1se used throughout is done and ready for add, commit.

**Comment by jaclyn-taroni** [comment]
@bill-amadio I made changes to your branch to fix conflicts and get things working in CI. I think since you've added your changes you will want to commit those, pull, and then push. There may be conflicts that arise so you could instead go the `stash` route: see [the answer with the most votes from this StackOverflow post](https://stackoverflow.com/questions/18529206/when-do-i-need-to-do-git-pull-before-or-after-git-add-git-commit). Feel free to pop in to Slack if you need help 😄

**Comment by bill-amadio** [comment]
Thanks for the merge and all your help, Jackie.  It's nice to see the file in the AlexsLemonade master.

Re the items not in the Notebook that should be reported, the modular design is effective.

Within 01-clean_split_data.R, I would filter using toupper(reported_gender) == "MALE" | toupper(reported_gender) == "FEMALE".  There were samples with reported_gender value "Not Available" in V5, so choosing what we want over what should be eliminated seems less likely to leave unusable samples.

02-train_elasticnet.R is clear.  I do not have anything to add here.

In 03-evaluate_model.R, I prefer the logic of doing all the saves here at the end of the process.  And saving the full histologies file feels like it would lend more flexibility to downstream scripts than saving a column of labels.

**Comment by jaclyn-taroni** [comment]
@bill-amadio the plan you've laid out sounds good to me.

**Comment by bill-amadio** [comment]
> 
> 
> @bill-amadio the plan you've laid out sounds good to me.

OK, @jaclyn-taroni.  I have tomorrow set aside to work on this.
